File: packages/python/tanstack-ai/src/tanstack_ai/base_adapter.py
"""
Base adapter class for AI providers.

All AI adapters should inherit from this class and implement the required methods.
"""

import time
from abc import ABC, abstractmethod
from typing import AsyncIterator, List

from .types import (
    AIAdapterConfig,
    ChatOptions,
    EmbeddingOptions,
    EmbeddingResult,
    StreamChunk,
    SummarizationOptions,
    SummarizationResult,
)


class BaseAdapter(ABC):
    """
    Base adapter class with support for endpoint-specific models and provider options.

    All concrete adapters (OpenAI, Anthropic, etc.) should inherit from this class
    and implement the abstract methods.
    """

    def __init__(self, config: AIAdapterConfig = AIAdapterConfig()):
        """
        Initialize the adapter.

        Args:
            config: Configuration for the adapter (API key, base URL, etc.)
        """
        self.config = config

    @property
    @abstractmethod
    def name(self) -> str:
        """Name of the adapter (e.g., 'openai', 'anthropic')."""
        pass

    @property
    @abstractmethod
    def models(self) -> List[str]:
        """List of models that support chat/text completion."""
        pass

    @property
    def embedding_models(self) -> List[str]:
        """List of models that support embeddings."""
        return []

    @abstractmethod
    async def chat_stream(self, options: ChatOptions) -> AsyncIterator[StreamChunk]:
        """
        Stream chat completions with structured JSON chunks.

        Args:
            options: Chat options including model, messages, tools, etc.

        Yields:
            StreamChunk objects (content, tool_call, tool_result, done, error, etc.)
        """
        pass

    @abstractmethod
    async def summarize(self, options: SummarizationOptions) -> SummarizationResult:
        """
        Summarize text.

        Args:
            options: Summarization options

        Returns:
            SummarizationResult with the summary and usage info
        """
        pass

    @abstractmethod
    async def create_embeddings(self, options: EmbeddingOptions) -> EmbeddingResult:
        """
        Create embeddings for text.

        Args:
            options: Embedding options

        Returns:
            EmbeddingResult with embeddings and usage info
        """
        pass

    def _generate_id(self) -> str:
        """Generate a unique ID for requests."""
        import random
        import string

        timestamp = int(time.time() * 1000)
        random_suffix = "".join(
            random.choices(string.ascii_lowercase + string.digits, k=7)
        )
        return f"{self.name}-{timestamp}-{random_suffix}"

File: packages/typescript/ai/src/base-adapter.ts
import type {
  AIAdapter,
  AIAdapterConfig,
  ChatOptions,
  DefaultMessageMetadataByModality,
  EmbeddingOptions,
  EmbeddingResult,
  Modality,
  StreamChunk,
  SummarizationOptions,
  SummarizationResult,
} from './types'

/**
 * Base adapter class with support for endpoint-specific models and provider options.
 *
 * Generic parameters:
 * - TChatModels: Models that support chat/text completion
 * - TEmbeddingModels: Models that support embeddings
 * - TChatProviderOptions: Provider-specific options for chat endpoint
 * - TEmbeddingProviderOptions: Provider-specific options for embedding endpoint
 * - TModelProviderOptionsByName: Provider-specific options for model by name
 * - TModelInputModalitiesByName: Map from model name to its supported input modalities
 * - TMessageMetadataByModality: Map from modality type to adapter-specific metadata types
 */
export abstract class BaseAdapter<
  TChatModels extends ReadonlyArray<string> = ReadonlyArray<string>,
  TEmbeddingModels extends ReadonlyArray<string> = ReadonlyArray<string>,
  TChatProviderOptions extends Record<string, any> = Record<string, any>,
  TEmbeddingProviderOptions extends Record<string, any> = Record<string, any>,
  TModelProviderOptionsByName extends Record<string, any> = Record<string, any>,
  TModelInputModalitiesByName extends Record<
    string,
    ReadonlyArray<Modality>
  > = Record<string, ReadonlyArray<Modality>>,
  TMessageMetadataByModality extends {
    image: unknown
    audio: unknown
    video: unknown
    document: unknown
  } = DefaultMessageMetadataByModality,
> implements
    AIAdapter<
      TChatModels,
      TEmbeddingModels,
      TChatProviderOptions,
      TEmbeddingProviderOptions,
      TModelProviderOptionsByName,
      TModelInputModalitiesByName,
      TMessageMetadataByModality
    >
{
  abstract name: string
  abstract models: TChatModels
  embeddingModels?: TEmbeddingModels
  protected config: AIAdapterConfig

  // These properties are used for type inference only, never assigned at runtime
  _providerOptions?: TChatProviderOptions
  _chatProviderOptions?: TChatProviderOptions
  _embeddingProviderOptions?: TEmbeddingProviderOptions
  // Type-only map; concrete adapters should override this with a precise type
  _modelProviderOptionsByName!: TModelProviderOptionsByName
  // Type-only map for model input modalities; concrete adapters should override this
  _modelInputModalitiesByName?: TModelInputModalitiesByName
  // Type-only map for message metadata types; concrete adapters should override this
  _messageMetadataByModality?: TMessageMetadataByModality

  constructor(config: AIAdapterConfig = {}) {
    this.config = config
  }

  abstract chatStream(options: ChatOptions): AsyncIterable<StreamChunk>

  abstract summarize(
    options: SummarizationOptions,
  ): Promise<SummarizationResult>
  abstract createEmbeddings(options: EmbeddingOptions): Promise<EmbeddingResult>

  protected generateId(): string {
    return `${this.name}-${Date.now()}-${Math.random()
      .toString(36)
      .substring(7)}`
  }
}
